#### https://stats.idre.ucla.edu/r/library/r-library-contrast-coding-systems-for-categorical-variables/
#### and https://leanpub.com/principlesoffmri
#### and https://stats.idre.ucla.edu/spss/faq/coding-systems-for-categorical-variables-in-regression-analysis-2/

## Contrasts

How can we estimate the signal magnitudes in response to a single condition, as an average over multiple conditions, or as a difference in magnitude across conditions?

Within the GLM framework using contrasts, or linear combinations of the $\beta$ values,
we can easily handle all these questions. For example, consider that we have two conditions, A and B,
which reflect periods of auditory and visual stimulation, respectively. The contrast [A - B] estimates
the activation difference for auditory - visual stimulation, which we can test by applying a linear
contrast vector c (i.e. calculating c
T β where c
T = [1 − 1]).

c is a contrast matrix with a row for each parameter or linear combination of parameters (e.g. a set of nuisance covariates)

##### Rules of Thumb

<ol>
    <li> If you test a difference between conditions, then contrast weights should always sum to zero. However,  if you test the average of one or more conditions against the implicit baseline, this need not hold
    <li> While scaling of weights affects magnitude (e.g. [1 -1] vs. [.5 -.5]), it does not influence inference (p-values, t-tests)
</ol>

## Loading Example Data for Race (4 levels)

In [5]:
hsb2 = read.table('https://stats.idre.ucla.edu/stat/data/hsb2.csv', header=T, sep=",")

In [6]:
hsb2

id  female race ses schtyp prog read write math science socst
1    70 0      4    1   1      1    57   52    41   47      57   
2   121 1      4    2   1      3    68   59    53   63      61   
3    86 0      4    3   1      1    44   33    54   58      31   
4   141 0      4    3   1      3    63   44    47   53      56   
5   172 0      4    2   1      2    47   52    57   53      61   
6   113 0      4    2   1      2    44   52    51   63      61   
7    50 0      3    2   1      1    50   59    42   53      61   
8    11 0      1    2   1      2    34   46    45   39      36   
9    84 0      4    2   1      1    63   57    54   58      51   
10   48 0      3    2   1      2    57   55    52   50      51   
11   75 0      4    2   1      3    60   46    51   53      61   
12   60 0      4    2   1      2    57   65    51   63      61   
13   95 0      4    3   1      2    73   60    71   61      71   
14  104 0      4    3   1      2    54   63    57   55      46   
15   38 0      3    1   1      2    45   57    50   31      56   
16  115 0      4    1   1      1    42   49    43   50      56   
17   76 0      4    3   1      2    47   52    51   50      56   
18  195 0      4    2   2      1    57   57    60   58      56   
19  114 0      4    3   1      2    68   65    62   55      61   
20   85 0      4    2   1      1    55   39    57   53      46   
21  167 0      4    2   1      1    63   49    35   66      41   
22  143 0      4    2   1      3    63   63    75   72      66   
23   41 0      3    2   1      2    50   40    45   55      56   
24   20 0      1    3   1      2    60   52    57   61      61   
25   12 0      1    2   1      3    37   44    45   39      46   
26   53 0      3    2   1      3    34   37    46   39      31   
27  154 0      4    3   1      2    65   65    66   61      66   
28  178 0      4    2   2      3    47   57    57   58      46   
29  196 0      4    3   2      2    44   38    49   39      46   
30   29 0      2    1   1      1    52   44    49   55      41   
⋮   ⋮   ⋮      ⋮    ⋮   ⋮      ⋮    ⋮    ⋮     ⋮    ⋮       ⋮    
171 119 1      4    1   1      1    42   57    45   50      43   
172  13 1      1    2   1      3    47   46    39   47      61   
173  51 1      3    3   1      1    42   36    42   31      39   
174  26 1      2    3   1      2    60   59    62   61      51   
175  36 1      3    1   1      1    44   49    44   35      51   
176 135 1      4    1   1      2    63   60    65   54      66   
177  59 1      4    2   1      2    65   67    63   55      71   
178  78 1      4    2   1      2    39   54    54   53      41   
179  64 1      4    3   1      3    50   52    45   58      36   
180  63 1      4    1   1      1    52   65    60   56      51   
181  79 1      4    2   1      2    60   62    49   50      51   
182 193 1      4    2   2      2    44   49    48   39      51   
183  92 1      4    3   1      1    52   67    57   63      61   
184 160 1      4    2   1      2    55   65    55   50      61   
185  32 1      2    3   1      3    50   67    66   66      56   
186  23 1      2    1   1      2    65   65    64   58      71   
187 158 1      4    2   1      1    52   54    55   53      51   
188  25 1      2    2   1      1    47   44    42   42      36   
189 188 1      4    3   2      2    63   62    56   55      61   
190  52 1      3    1   1      2    50   46    53   53      66   
191 124 1      4    1   1      3    42   54    41   42      41   
192 175 1      4    3   2      1    36   57    42   50      41   
193 184 1      4    2   2      3    50   52    53   55      56   
194  30 1      2    3   1      2    41   59    42   34      51   
195 179 1      4    2   2      2    47   65    60   50      56   
196  31 1      2    2   2      1    55   59    52   42      56   
197 145 1      4    2   1      3    42   46    38   36      46   
198 187 1      4    2   2      1    57   41    57   55      52   
199 118 1      4    2   1      1    55   62 

In [7]:
#creating the factor variable race.f
hsb2$race.f = factor(hsb2$race, labels=c("Hispanic", "Asian", "African-Am", "Caucasian"))

#Before considering any analyses, let’s look at the mean of the dependent variable, write, 
#for each level of race.  This will help in interpreting the output from later analyses.

In [8]:
tapply(hsb2$write, hsb2$race.f, mean)

ERROR while rich displaying an object: Error in dn[[2L]]: subscript out of bounds

Traceback:
1. FUN(X[[i]], ...)
2. tryCatch(withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
3. tryCatchList(expr, classes, parentenv, handlers)
4. tryCatchOne(expr, names, parentenv, handlers[[1L]])
5. doTryCatch(return(expr), name, parentenv, handler)
6. withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler)
7. mime2repr[[mime]](obj)
8. repr_markdown.numeric(obj)
9. repr_vector_generic(html_escape_names(obj), "%s. %s\n", "%s\n:   %s", 
 .     "**%s:** %s", "%s\n\n", item_uses_numbers = TRUE, escape_fun = html_escape)
10. html_escape_names(obj)
11. .escape_names(obj, "html")
12. colnames(obj)
ERROR while rich displaying an object: 

Hispanic      Asian African-Am  Caucasian 
  46.45833   58.00000   48.20000   54.05517

## 1. Dummy Coding

In [10]:
#the contrast matrix for categorical variable with four levels
contr.treatment(4)

2 3 4
1 0 0 0
2 1 0 0
3 0 1 0
4 0 0 1

Dummy coding is probably the most commonly used coding scheme. It compares each level of the categorical variable to a fixed reference level. For example, we can choose race = 1 as the reference group and compare the mean of variable write for each level of race 2, 3 and 4 to the reference level of 1. This is the default for unordered factors in R.

In [8]:
#assigning the treatment contrasts to race.f
contrasts(hsb2$race.f) = contr.treatment(4)
#the regression
summary(lm(write ~ race.f, hsb2))


Call:
lm(formula = write ~ race.f, data = hsb2)

Residuals:
     Min       1Q   Median       3Q      Max 
-23.0552  -5.4583   0.9724   7.0000  18.8000 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   46.458      1.842  25.218  < 2e-16 ***
race.f2       11.542      3.286   3.512 0.000552 ***
race.f3        1.742      2.732   0.637 0.524613    
race.f4        7.597      1.989   3.820 0.000179 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.025 on 196 degrees of freedom
Multiple R-squared:  0.1071,	Adjusted R-squared:  0.0934 
F-statistic: 7.833 on 3 and 196 DF,  p-value: 5.785e-05


The parameter estimate for the first contrast compares the mean of the dependent variable, write, for levels 1 and 2 yielding 11.5417 and is statistically significant (p<.000). The t-value associated with this test is 3.5122.  The results of the second contrast, comparing the mean of write for levels 1 and 3. The expected difference in variable write between group 1 and 3 is 1.7417 and  is not statistically significant (t = 0.6374, p = .5246), while the third contrast is statistically significant. Notice that the intercept corresponds to the cell mean for race = Hispanic group.

### 2. Deviation Coding

This coding system compares the mean of the dependent variable for a given level to the overall mean of the dependent variable.  In our example below, the first comparison compares level 1 (Hispanics) to all levels of race, the second comparison compares level 2 (Asians) to all levels of race, and the third comparison compares level 3 (African Americans) to all levels of race.

As you see in the example below, the regression coding is accomplished by assigning 1 to level 1 for the first comparison (because level 1 is the level to be compared to all others), a 1 to level 2 for the second comparison (because level 2 is to be compared to all others), and 1 to level 3 for the third comparison (because level 3 is to be compared to all others).  Note that a  -1 is assigned to level 4 for all three comparisons (because it is the level that is never compared to the other levels) and all other values are assigned a 0.  This regression coding scheme yields the comparisons described above.
We will not create the contrast matrix manually because the contr.sum function creates it for us.

In [2]:
#the contrast matrix for categorical variable with four levels
contr.sum(4)

[,1] [,2] [,3]
1  1    0    0  
2  0    1    0  
3  0    0    1  
4 -1   -1   -1

In [9]:
#assigning the deviation contrasts to race.f
contrasts(hsb2$race.f) = contr.sum(4)
#the regression
summary(lm(write ~ race.f, hsb2))


Call:
lm(formula = write ~ race.f, data = hsb2)

Residuals:
     Min       1Q   Median       3Q      Max 
-23.0552  -5.4583   0.9724   7.0000  18.8000 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  51.6784     0.9821  52.619  < 2e-16 ***
race.f1      -5.2200     1.6314  -3.200  0.00160 ** 
race.f2       6.3216     2.1603   2.926  0.00384 ** 
race.f3      -3.4784     1.7323  -2.008  0.04602 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.025 on 196 degrees of freedom
Multiple R-squared:  0.1071,	Adjusted R-squared:  0.0934 
F-statistic: 7.833 on 3 and 196 DF,  p-value: 5.785e-05


The contrast estimate is the mean for level 1 minus the grand mean.  However, this grand mean is not the mean of the dependent variable that is listed in the output of the means command above.  Rather it is the mean of means of the dependent variable at each level of the categorical variable:  (46.4583 + 58 + 48.2 + 54.0552) / 4 = 51.678375.  This contrast estimate is then 46.4583 – 51.678375 = -5.220.  The difference between this value and zero (the null hypothesis that the contrast coefficient is zero) is statistically significant (p = .0016), and the t-value for this test of -3.20. The results for the next two contrasts were computed in a similar manner.

### 3. Helmert Coding

Helmert coding compares each level of a categorical variable to the mean of the subsequent levels.  Hence, the first contrast compares the mean of the dependent variable for level 1 of race with the mean of all of the subsequent levels of race (levels 2, 3, and 4), the second contrast compares the mean of the dependent variable for level 2 of race with the mean of all of the subsequent levels of race (levels 3 and 4), and the third contrast compares the mean of the dependent variable for level 3 of race with the mean of all of the subsequent levels of race (level 4).

In R there is a built-in function, contr.helmert, which, up to a scale,  corresponds to the reverse Helmert coding discussed in next section.  We will create the contrast matrix for Helmert coding manually below.

Below we see an example of Helmert regression coding.  For the first comparison (comparing level 1 with levels 2, 3 and 4) the codes are 3/4 and -1/4 -1/4 -1/4.  The second comparison compares level 2 with levels 3 and 4 and is coded 0 2/3 -1/3 -1/3.  The third comparison compares level 3 to level 4 and is coded 0 0 1/2 -1/2.

In [10]:
#helmert for factor variable with 4 levels
my.helmert = matrix(c(3/4, -1/4, -1/4, -1/4, 0, 2/3, -1/3, -1/3, 0, 0, 1/
	2, -1/2), ncol = 3)
my.helmert

[,1]  [,2]       [,3]
[1,]  0.75  0.0000000  0.0
[2,] -0.25  0.6666667  0.0
[3,] -0.25 -0.3333333  0.5
[4,] -0.25 -0.3333333 -0.5

In [11]:
#assigning the new helmert coding to race.f
contrasts(hsb2$race.f) = my.helmert
#the regression
summary(lm(write ~ race.f, hsb2))


Call:
lm(formula = write ~ race.f, data = hsb2)

Residuals:
     Min       1Q   Median       3Q      Max 
-23.0552  -5.4583   0.9724   7.0000  18.8000 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  51.6784     0.9821  52.619  < 2e-16 ***
race.f1      -6.9601     2.1752  -3.200  0.00160 ** 
race.f2       6.8724     2.9263   2.348  0.01985 *  
race.f3      -5.8552     2.1528  -2.720  0.00712 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 9.025 on 196 degrees of freedom
Multiple R-squared:  0.1071,	Adjusted R-squared:  0.0934 
F-statistic: 7.833 on 3 and 196 DF,  p-value: 5.785e-05


The contrast estimate for the comparison between level 1 and the remaining levels is calculated by taking the mean of the dependent variable for level 1 and subtracting the mean of the dependent variable for levels 2, 3 and 4: 46.4583 – [(58 + 48.2 + 54.0552) / 3] = -6.960, which is statistically significant.  This means that the mean of write for level 1 of race is statistically significantly different from the mean of write for levels 2 through 4. To calculate the contrast coefficient for the comparison between level 2 and the later levels, you subtract the mean of the dependent variable for levels 3 and 4 from the mean of the dependent variable for level 2:  58 – [(48.2 + 54.0552) / 2] = 6.872, which is statistically significant.  The contrast estimate for the comparison between level 3 and level 4 is the difference between the mean of the dependent variable for the two levels:  48.2 – 54.0552 = -5.855, which is also statistically significant.

### 4. Orthogonal Polynomial

Orthogonal polynomial coding is a form of trend analysis in that it is looking for the linear, quadratic and cubic trends in the categorical variable. This type of coding system should be used only with an ordinal variable in which the levels are equally spaced. Examples of such a variable might be income or education. The table below shows the contrast coefficients for the linear, quadratic and cubic trends for the four levels. In R it is not necessary to compute these values since this contrast can be obtained for any categorical variable by using the contr.poly function. This is also the default contrast used for ordered factor variables.

For the purpose of illustration, let’s create an ordered categorical variable based on the variable read. Notice that this is the only example on this page that does not use race as the categorical variable, since it is not ordered.

In [18]:
#using read values because race is not ordinal
hsb2$readcat<-cut(hsb2$read, 4, ordered = TRUE)
table(hsb2$readcat)


(28,40] (40,52] (52,64] (64,76] 
     22      93      55      30 

In [19]:
tapply(hsb2$write, hsb2$readcat, mean)

ERROR while rich displaying an object: Error in dn[[2L]]: subscript out of bounds

Traceback:
1. FUN(X[[i]], ...)
2. tryCatch(withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
3. tryCatchList(expr, classes, parentenv, handlers)
4. tryCatchOne(expr, names, parentenv, handlers[[1L]])
5. doTryCatch(return(expr), name, parentenv, handler)
6. withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler)
7. mime2repr[[mime]](obj)
8. repr_markdown.numeric(obj)
9. repr_vector_generic(html_escape_names(obj), "%s. %s\n", "%s\n:   %s", 
 .     "**%s:** %s", "%s\n\n", item_uses_numbers = TRUE, escape_fun = html_escape)
10. html_escape_names(obj)
11. .escape_names(obj, "html")
12. colnames(obj)
ERROR while rich displaying an object: 

(28,40]  (40,52]  (52,64]  (64,76] 
42.77273 49.97849 56.56364 61.83333

In [20]:
#the contrast matrix for categorical variable with four levels
contr.poly(4)

.L         .Q   .C        
[1,] -0.6708204  0.5 -0.2236068
[2,] -0.2236068 -0.5  0.6708204
[3,]  0.2236068 -0.5 -0.6708204
[4,]  0.6708204  0.5  0.2236068

In [21]:
#assigning the orthogonal polynomial contrasts to readcat
contrasts(hsb2$readcat) = contr.poly(4)
summary(lm(write ~ readcat, hsb2))


Call:
lm(formula = write ~ readcat, data = hsb2)

Residuals:
    Min      1Q  Median      3Q     Max 
-18.979  -5.824   1.227   5.436  17.021 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  52.7870     0.6339  83.268   <2e-16 ***
readcat.L    14.2587     1.4841   9.607   <2e-16 ***
readcat.Q    -0.9680     1.2679  -0.764    0.446    
readcat.C    -0.1554     1.0062  -0.154    0.877    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 7.726 on 196 degrees of freedom
Multiple R-squared:  0.3456,	Adjusted R-squared:  0.3356 
F-statistic: 34.51 on 3 and 196 DF,  p-value: < 2.2e-16


The regression results indicate a strong linear effect of readcat on the outcome variable write. There is not a significant quadratic effect nor a cubic effect of readcat on the outcome variable write.